nearswap/src/twap.rs

```diff
@@ -0,0 +1,576 @@
+        if self.observations.len == 0 {
+            self.current_idx = self.initialize(env::block_timestamp(), price1, price2);
+        } else {
+            self.current_idx = self.write(
+                env::block_timestamp(),
+                price1,
+                price2,
+                MAX_LENGTH
+            );
+        }
+
+        // update mean
+        let len: usize;
+        if self.pivoted == true {
+            len = MAX_LENGTH.into();
+        } else {
+            len = self.observations.len();
+        }
+
+        self.update_mean(len);
+    }
+
+    pub fn update_mean(&mut self, len: usize) {
+        self.mean_1min = self.calculate_mean(Mean::M_1MIN, len);
+
+        self.mean_5min = self.calculate_mean(Mean::M_5MIN, len);
+
+        self.mean_1h = self.calculate_mean(Mean::M_1H, len);
+
+        self.mean_12h = self.calculate_mean(Mean::M_12H, len);
+    }
+}
+
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct Observation {
+    // timestamp
+    pub block_timestamp: u64,
+    // Number of observations till block_timestamp
+    pub num_of_observations: u128,
+    // cumulative price of token1 till block_timestamp
+    pub price1_cumulative: u128,
+    // cumulative price of token2 till block_timestamp
+    pub price2_cumulative: u128,
+}
+
+impl Observation {
+    /**
+    Transforms a previous observation into a new observation.
+    Parameters:
+    + `block_timestamp`: _must_ be chronologically equal to or greater than last.block_timestamp.
+    + `last`: The specified observation to be transformed.
+    + `price1`: price of first token.
+    + `price2`: price of second token.
+    */
+    pub fn transform(
+        last: &Observation,
+        block_timestamp: u64,
+        price1: u128,
+        price2: u128
+    ) -> Observation {
+        return
+            Observation {
+                block_timestamp: block_timestamp,
+                num_of_observations: last.num_of_observations + 1,
+                price1_cumulative: last.price1_cumulative + price1,
+                price2_cumulative: last.price2_cumulative + price2,
+            };
+    }
+
+    /// returns instance of Observation structure
+    pub fn new() -> Self {
+        return Self {
+            block_timestamp: 1,
+            num_of_observations: 0,
+            price1_cumulative: 0,
+            price2_cumulative: 0,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::Twap;
+    use super::Observation;
+    use super::*;
+
+    use near_sdk::test_utils::{accounts, VMContextBuilder};
+    use near_sdk::{testing_env, BlockHeight, MockedBlockchain};
+
+    fn init_blockchain() {
+        let context = VMContextBuilder::new();
+        testing_env!(context.build());
+    }
+
+    // returns twap with observation vector with timestamp [1, 2, 3, 4, 5, 6, 7, 9, 10]
+    fn get_twap() -> Twap {
+
+        let mut twap: Twap = Twap::new();
+        let mut current_idx = twap.initialize(env::block_timestamp(), 1, 1);
```

Comment:
instead of `env::block_timestamp()` you can put some number or an argument to the `get_twap`.

Reply:
![up](https://user-images.githubusercontent.com/26249903/114503169-7fc40d80-9c4a-11eb-8a70-9b7dde318150.png)

tried this.

Reply:
I meant to have:
```
fn get_twap(block_timestamp: u64) -> Twap {...
```
So, to not use any `env::` function in this file.

Reply:
ok, but we still have in log_observation, this can be removed if we pass blocktimestamp from pool.rs. please let me know if you think i should do that

Reply:
yes, this should come as a function argument. We shouldn't use any `env::*` function here, however we can sue constants.